### ISSUE TITLE
Run API StringIO TextReporter not capturing output stream with --output-format=json

### ISSUE THREAD

**Issue by RobFirth** [PyCQA/pylint#8164]
### Bug description

Running the following MWE does not return the report in the expected format (json), instead prints the json formatted report to `stdout` rather than capturing the stream and returning it in the reporter, as per the docs [here](https://pylint.pycqa.org/en/v2.16.0/development_guide/api/pylint.html#recover-the-result-in-a-stream).

```python
from io import StringIO

from pylint.lint import Run
from pylint.reporters.text import TextReporter

pylint_output = StringIO()  # Custom open stream
reporter = TextReporter(pylint_output)

Run(args=["pylint_debug_target.py", "--output-format=json"], reporter=reporter, exit=False)

report = pylint_output.getvalue()
print(report)  # Retrieve and print the text report
```

where `pylint_debug_target.py` is:

```python
"""
this file should make pylint throw an error (E0102) 
"""

def hello_zoidberg():
    """Need a MWE?"""
    print("Why not Zoidberg?")

def pylint_e0102():
    """First definition of the function"""
    print("foo")

def pylint_e0102():
    """Second definition of the function"""
    print("bar")

if __name__ == "__main__":
    hello_zoidberg()
else:
    pass
```

### Default formatted output

```python
from io import StringIO

from pylint.lint import Run
from pylint.reporters.text import TextReporter

pylint_output = StringIO()  # Custom open stream
reporter = TextReporter(pylint_output)

Run(["pylint_debug_target.py"], reporter=reporter, exit=False)

report = pylint_output.getvalue()
print(report)  # Retrieve and print the text report
```

returns the report as expected as:

```
[
    {
        "type": "error",
        "module": "pylint_debug_target",
        "obj": "pylint_e0102",
        "line": 13,
        "column": 0,
        "endLine": 13,
        "endColumn": 16,
        "path": "pylint_debug_target.py",
        "symbol": "function-redefined",
        "message": "function already defined line 9",
        "message-id": "E0102"
    }
]
```

### Note
The print to `stdout` occurs [here](https://github.com/PyCQA/pylint/blob/main/pylint/lint/run.py#L213), so maybe the `_output` method is not being set correctly in the `Run` configuration?

This is an issue caused by migrating from `epylint`, where the `stdout` and `stderr` were returned directly.

* output options detailed here - https://pylint.pycqa.org/en/v2.16.0/user_guide/usage/output.html
* following API docs here - https://pylint.pycqa.org/en/v2.16.0/development_guide/api/pylint.html#recover-the-result-in-a-stream

### Command used

```shell
from io import StringIO

from pylint.lint import Run
from pylint.reporters.text import TextReporter

pylint_output = StringIO()  # Custom open stream
reporter = TextReporter(pylint_output)

Run(args=["pylint_debug_target.py", "--output-format=json"], reporter=reporter, exit=False)
```

### Pylint output

```shell
to stdout:

[
    {
        "type": "error",
        "module": "pylint_debug_target",
        "obj": "pylint_e0102",
        "line": 13,
        "column": 0,
        "endLine": 13,
        "endColumn": 16,
        "path": "pylint_debug_target.py",
        "symbol": "function-redefined",
        "message": "function already defined line 9",
        "message-id": "E0102"
    }
]
```

but `pylint_output.getvalue() == None` from reporter.

### Expected behavior

```python
from io import StringIO

from pylint.lint import Run
from pylint.reporters.text import TextReporter

pylint_output = StringIO()  # Custom open stream
reporter = TextReporter(pylint_output)

Run(args=["pylint_debug_target.py", "--output-format=json"], reporter=reporter, exit=False)
```

gives:

```python
pylint_output.getvalue() == [
    {
        "type": "error",
        "module": "pylint_debug_target",
        "obj": "pylint_e0102",
        "line": 13,
        "column": 0,
        "endLine": 13,
        "endColumn": 16,
        "path": "pylint_debug_target.py",
        "symbol": "function-redefined",
        "message": "function already defined line 9",
        "message-id": "E0102"
    }
]
```

### Pylint version

```shell
pylint 2.16.0
astroid 2.14.1
Python 3.10.6 (main, Nov 14 2022, 14:18:33) [Clang 13.1.6 (clang-1316.0.21.2.5)]
```

### OS / Environment

MacOS Monterey 12.6.2

### Additional dependencies

astroid==2.14.1
dill==0.3.6
isort==5.12.0
lazy-object-proxy==1.9.0
mccabe==0.7.0
platformdirs==2.6.2
pylint==2.16.0
tomli==2.0.1
tomlkit==0.11.6
typing_extensions==4.4.0
wrapt==1.14.1

**Comment by DanielNoord** [comment]
I'm not sure if ``--output-format`` and `reporter` are supposed to work together. Did this work previously?

**Comment by RobFirth** [comment]
> I'm not sure if ``--output-format`` and `reporter` are supposed to work together. Did this work previously?

We weren't using the `Run` API previously - we were capturing the STDOUT from the run methods in `epylint`, which are now unsupported.

I did wonder though!

**Comment by DanielNoord** [comment]
I think this is just an incompatibility. Can't you import the `JSONReporter` and then skip the ``--output-format``?

**Comment by RobFirth** [comment]
> I think this is just an incompatibility. Can't you import the `JSONReporter` and then skip the ``--output-format``?

Thanks for the suggestion - I will give this a try.

**Comment by RobFirth** [comment]
@DanielNoord I got the expected behaviour with the `JSONReporter`, thanks for the pointer.

```python
## JSONReporter

from io import StringIO

from pylint.lint import Run
from pylint.reporters import JSONReporter

pylint_output = StringIO()  # Custom open stream
reporter = JSONReporter(pylint_output)

Run(args=["pylint_debug_target.py"], reporter=reporter, exit=False)
# Run("pylint_debug_target.py", args=["--rcfile=pyproject.toml", "--output-format=json"], reporter=reporter, exit=False)

report = pylint_output.getvalue()
print(report)  # Retrieve and print the text report
```

**Comment by DanielNoord** [comment]
Closing this issue then! Thanks for getting back to us!
